File: exercises/js_challenges/CtCI/chatper_2_linked_lists/2.2_return_kth_to_last.js
// Implement an algorithm to find the kth to last element of a
// singly linked list.

function Node(data, next = nil){
  this.data = data;
  this.next = next;
}

let list = new Node(5)
list = new Node(4, list)
list = new Node(3, list)
list = new Node(10, list)
list = new Node(3, list)

function kth_last(head, n){
  for(let i = 0; i < n-1; i++){
    if (head.next == null){
      return null
    }
    head = head.next
  }
  return head
}

console.log(kth_last(list,2))

File: exercises/ruby_challenges/CtCI/chapter_2_linked_lists/2.2_return_kth_to_last.rb
# Implement an algorithm to find the kth to last element of a
# singly linked list.

class Node
attr_accessor :data, :next
  def initalizae(data,next_node=nil)
    @data = data
    @next = next_node
  end
end

list = Node.new(3)
list = Node.new(4, list)
list = Node.new(5, list)
list = Node.new(10,list)
list = Node.new(3, list)

def kth_last(head,n)
  i = 0
  while(i < n-1)
    if(head == nil || head.next == nil)
      return nil
    end
    head = head.next
    i+=1
  end
  return head
end

puts kth_last(list,2)